docs/proposals/custom-metrics-api.md

```diff
@@ -0,0 +1,189 @@
+  namespace.
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}?labelSelector=foo`:
+  retrieve the given metric for all objects of the given type matching the
+  given label selector in the given namespace.
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}?names=foo,bar`:
+  retrieve the given metric for the objects of the given type with the
+  given name in the given namespace
+
+For example, to retrieve the custom metric "hits-per-second" for all pods
+matching "app=frontend` in the namespaces "webapp", the request might look
+like:
+
+`/apis/general-metrics/v1alpha1/namespaces/webapp/object-metrics/pods/hits-per-second?labelSelector=app%3Dfrontend`.
+
+API Objects
+-----------
+
+The request URLs listed above will return objects of the `Metrics`
+type, described below:
+
+```go
+
+// a list of values for a given metric for some set of objects
+type Metrics struct {
+    unversioned.TypeMeta `json:",inline"`
+    unversioned.ListMeta `json:"metadata,omitempty"`
+
+    // the name of the metric
+    MetricName string `json:"metricName"`
+
+    // the value of the metric across the described objects
+    MetricValues []MetricValue `json:"metricValues"`
+}
+
+// a metric value for some object
+type GenericMetrics struct {
+    // the type and object metadata describes the object in question
+    unversioned.TypeMeta `json:",inline"`
+    v1.ObjectMeta        `json:"metadata,omitempty"`
+
+    // indicates the end of the time window containing these metrics (i.e.
+    // these metrics come from some time in [Timestamp-Window, Timestamp])
+    Timestamp unversioned.Time `json:"timestamp"`
+
+    // indicates the duration of the time window containing these metrics
+    Window    unversioned.Duration `json:"window"`
+
+    // the value of the metric for this
+    Value float64
+}
+```
+
+For instance, the example request above would yield the following object:
+
+```json
+{
+    "kind": "Metrics",
+    "apiVersion": "general-metrics/v1alpha1",
+    "metricName": "hits-per-second",
+    "metricValues": [
+        {
+            "kind": "Pod",
+            "metadata": { "name": "server1", "namespace": "webapp" },
+            "timestamp": SOME_TIMESTAMP_HERE,
+            "window": "10s",
+            "value": 10.0
+        },
+        {
+            "kind": "Pod",
+            "metadata": { "name": "server2", "namespace": "webapp" },
+            "timestamp": SOME_TIMESTAMP_HERE,
+            "window": "10s",
+            "value": 15.0
+        }
+    ]
+}
+```
+
+Mechanical Concerns
+-------------------
+
+This API is intended to be implemented by monitoring pipelines (e.g.
+inside Heapster, or as an adapter on top of a solution like Prometheus).
+It shares many mechanical requirements with normal Kubernetes APIs.
+
+The API should support encoding the result in both JSON and Protobuf,
+similarly to how core Kubernetes APIs work.  Most likely, the Protobuf
```

Comment:
I don't think we need Protobuffs any time soon. I think that the latency in monitoring pipeline itself will be large enough that we will not care that much about performance of communication protocol. This could become a problem in huge clusters where everything is autoscaled, but we should start with something simpler IMO.